AArch64: Add support for recoverable faults

carlocaione opened the pull request:
This patchset looks way bigger than it really is. It's basically addressing three things:
1. Only print `FAR_ELn` when needed (and prettify a bit the error output)
2. Add support for recoverable faults
3. Introduce CONFIG_EXCEPTION_DEBUG

carlocaione committed:
arch: aarch64: Branch SError vector table entry

Each vector table entry has 128-bytes to host the vector code. This is
not always enough and in general it's better to branch to the actual
exception handler elsewhere in memory.

Move the SError entry to a branched code.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: error: Beautify error printing

Make the printing of errors a bit more descriptive and print the FAR_ELn
register only when strictly required.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
aarch64: error: Enable recoverable errors

For some kind of faults we want to be able to put in action some
corrective actions and keep executing the code.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

andrewboie commented on arch/arm/core/aarch64/fatal.c:
```diff
@@ -179,11 +194,18 @@ void z_arm64_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
 			LOG_ERR("ELR_ELn: 0x%016llx", elr);
 
-			print_EC_cause(esr);
+			dump_esr(esf, esr, &dump_far, &recoverable);
```
it might be slightly better to separate the logic which dumps information, from the logic which determines whether the exception is recoverable.

I am not sure if in the current code here, whether gc-sections will discard all the strings if LOG is turned off. On x86 I had to ifdef around CONFIG_EXCEPTION_DEBUG to make sure it was.

carlocaione replied:
> it might be slightly better to separate the logic which dumps information, from the logic which determines whether the exception is recoverable.

Ok

> I am not sure if in the current code here, whether gc-sections will discard all the strings if LOG is turned off. On x86 I had to ifdef around CONFIG_EXCEPTION_DEBUG to make sure it was.

Indeed the strings are not discarded when `CONFIG_LOG=n`. I'll add a `CONFIG_EXCEPTION_DEBUG` for AArch64 as well.

tyrantlizard77 reviewed (approved):
Broadly speaking LGTM. I would check that the error information strings aren't still included in the binary if logging is turned off.

carlocaione committed:
aarch64: error: Introduce CONFIG_EXCEPTION_DEBUG

Introduce CONFIG_EXCEPTION_DEBUG to discard exception debug strings and
code when not needed.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>
